### PR TITLE
Tests, Comments, and minor code changes for mknod syscall

### PR DESCRIPTION
**Pull request by namanlalitnyu**
## Description

Fixes # (issue)
The following changes include the tests and comments in the code for the "mknod_syscall" file system call under RustPosix.
The tests were added to cover all the possible scenarios that might happen when calling the file system_call `mknod_syscall`. 

### Type of change
- [ ]  This change just contains the tests for an existing file system call.
- [ ]  This change contains the minor code changes and comments for mknod syscall.

## How Has This Been Tested?
Inorder to run the tests, we need to run `cargo test --lib` command inside the `safeposix-rust` directory.

All the tests are present under this directory: `lind_project/src/safeposix-rust/src/tests/fs_tests.rs`

- Test A - `ut_lind_fs_mknod_empty_path()`
- Test B - `ut_lind_fs_mknod_nonexisting_parent_directory()`
- Test C - `ut_lind_fs_mknod_existing_file()`
- Test D - `ut_lind_fs_mknod_invalid_modebits()`
- Test E -  `ut_lind_fs_mknod_invalid_filetypes()`
- Test F - `ut_lind_fs_mknod_success()`

## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

**Comment by namanlalitnyu** [comment]
I have a small doubt, as in @yashaswi2000's PR, I see that we have added a couple of lines for each test, which might be for setup. So, in order to add them here, should we wait for that PR to merge and then update this PR or add those changes here itself? 
Code changes: `acquiring a lock on TESTMUTEX prevents other tests from running concurrently
  // let mut _thelock = TESTMUTEX.lock().unwrap();
  //creates the file system if it doesnt exist yet.
  // setup::test_setup();`
cc: @JustinCappos @rennergade

File: src/safeposix/syscalls/fs_calls.rs
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs]: hmm I see how the original comment is incorrect but the way you changed it, i think the better comment is "Return an error if the provided path is empty."

**Comment by rennergade** [comment]
> I have a small doubt, as in @yashaswi2000's PR, I see that we have added a couple of lines for each test, which might be for setup. So, in order to add them here, should we wait for that PR to merge and then update this PR or add those changes here itself? Code changes: `acquiring a lock on TESTMUTEX prevents other tests from running concurrently // let mut _thelock = TESTMUTEX.lock().unwrap(); //creates the file system if it doesnt exist yet. // setup::test_setup();` cc: @JustinCappos @rennergade

Good question but I wouldn't worry about it for now. Yash will fix it up depending on the merge ordering.

**Comment by rennergade** [comment]
@Anway-Agte Can you give this a review before I do a full pass on it?

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: Don't put more than 4 spaces in a rustdoc comment.  This will cause rustdoc to think it's a test to run.

FYI: @yashaswi2000

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: wrap this and other comments to be <80 chars wide

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: Are you undoing `cargo fmt`?

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: Makes sense to me.  In general mentioning this in your issue / PR text is preferred.

File: src/tests/fs_tests.rs
Comment (JustinCappos) [inline src/tests/fs_tests.rs:36]: I think these may move to the new test format, but I'll let @yashaswi2000 weigh in...

**Review by JustinCappos — CHANGES_REQUESTED** [review]
A few nits.  Ping me when you've fixed them and resolved the conversations...

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated it!

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Fixed.

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Done

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Sorry, I missed this change. Have reverted it. 
I found that a few other formatting changes in a couple of files, including this one, need to be updated according to the "Rust format". We can probably have a small bug for it to get them fixed together.

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Sure.

File: src/tests/fs_tests.rs
Comment (namanlalitnyu) [reply src/tests/fs_tests.rs:36]: Closing this conversation based on the discussion today. Leaving the tests in this format as of now for this PR.

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [inline src/safeposix/syscalls/fs_calls.rs]: please add a Panics section too, which will list all the cases this can panic

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [inline src/safeposix/syscalls/fs_calls.rs]: is S_IFCHR part of the filetype bitmask? if so can be add a comment mentioning that

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [inline src/safeposix/syscalls/fs_calls.rs]: nit pick, if possible can we make all the syscall_errors in the three line format. its cool if not too

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Added the Panics section but currently we don't have any panics for this syscall.

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated the comment section.

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Fixed it.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs:506]: Why use the real time?  (I'm not complaining, just curious)

**Review by JustinCappos — APPROVED** [review]
Looks good!  One minor comment

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs:506]: Some programs in our code require logical times for execution, so we use real time for it. Also, Nginx has an assert condition, which requires time to be moving forward. 
Alternatively, we can use `clock counter` for it.

File: src/safeposix/syscalls/fs_calls.rs
Comment (Anway-Agte) [inline src/safeposix/syscalls/fs_calls.rs:441]: What does it mean for the mode bits to not be sane?

File: src/tests/fs_tests.rs
Comment (Anway-Agte) [inline src/tests/fs_tests.rs]: Shouldn't the test cases include mknod on different combinations of read/write/create permissions?

File: src/safeposix/syscalls/fs_calls.rs
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs:441]: It means to check if the mode bits, which are passed as function arguments are correctly set or not. If the mode bits are correctly set, then we should be able to create a new file with required file permissions which are represented by the "mode" bits. Otherwise, we can't create a file and return an error.

File: src/tests/fs_tests.rs
Comment (namanlalitnyu) [reply src/tests/fs_tests.rs]: Added the cases for different combinations. Actually, the `S_IRWXA` flag is the best possible combination for all the permissions, which covers all cases.

**Review by Anway-Agte — APPROVED** [review]

**Comment by rennergade** [comment]
Approved!

**Pull request merged**
